From 9d43cb7fdfad62781ca69e0ef89f7cb87e6e5ea5 Mon Sep 17 00:00:00 2001
From: Brian Dominy <brian@oddchange.com>
Date: Thu, 24 Jun 2010 22:49:20 -0400
Subject: [PATCH 13/37] Fix read() and write() system call return codes.

They should return number of bytes read/written on success, not 0.
---
 newlib/libc/sys/m6809sim/read.c  | 17 +++++++++--------
 newlib/libc/sys/m6809sim/write.c | 17 +++++++++--------
 2 files changed, 18 insertions(+), 16 deletions(-)

diff --git a/newlib/libc/sys/m6809sim/read.c b/newlib/libc/sys/m6809sim/read.c
index 314a6df..fa7d36e 100644
--- a/newlib/libc/sys/m6809sim/read.c
+++ b/newlib/libc/sys/m6809sim/read.c
@@ -15,16 +15,17 @@ inline unsigned char inbyte (void)
  */
 int read (int fd, char *buf, int len)
 {
-	if (fd == 0)
-	{
-		while (len-- > 0)
-			*buf++ = inbyte ();
-		return (0);
-	}
-	else
-	{
+	int _len = len;
+
+	if (fd != 0)
 		return (-1);
+
+	while (len-- > 0)
+	{
+		*buf++ = inbyte ();
+		_len--;
 	}
+	return len;
 }
 
 
diff --git a/newlib/libc/sys/m6809sim/write.c b/newlib/libc/sys/m6809sim/write.c
index e034791..4e669e0 100644
--- a/newlib/libc/sys/m6809sim/write.c
+++ b/newlib/libc/sys/m6809sim/write.c
@@ -16,16 +16,17 @@ inline void outbyte (unsigned char c)
  */
 int write (int fd, const char *buf, int len)
 {
-	if ((fd == 1) || (fd == 2))
-	{
-		while (len-- > 0)
-			outbyte (*buf++);
-		return (0);
-	}
-	else
-	{
+	int _len = len;
+
+	if (fd != 1 && fd != 2)
 		return (-1);
+
+	while (_len > 0)
+	{
+		outbyte (*buf++);
+		_len--;
 	}
+	return len;
 }
 
 
-- 
2.19.1.windows.1

